scripts/get-all-images.py

```diff
@@ -0,0 +1,223 @@
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def cleanup_images(images: list[str]) -> list[str]:
+    """
+    Given a list of OCI registry images ensure
+    1. there are no duplicates
+    2. there are no images with empty name
+    3. the list is sorted
+
+    Args:
+        images: List of images to be processed
+
+    Returns:
+        A new list without duplicate values, empty values and is sorted
+    """
+    images_set = set(images)
+    if '' in images_set:
+        images_set.remove('')
+
+    unique_images = list(images_set)
+    unique_images.sort()
+
+    return unique_images
+
+
+def get_bundle_images(bundle_path: str) -> list[str]:
+    """Return a list of images used by a bundle"""
+    bundle_dict = {}
+    with open(bundle_path, 'r') as file:
+        bundle_dict = yaml.safe_load(file)
+
+    images = []
+
+    for app_name, app in bundle_dict["applications"].items():
+        logging.info(f"Handling app {app_name}")
+
+        # exclude repos we know don't have images
+        if app_name in EXCLUDE_CHARMS:
+            logging.info("Ignoring charm %s", app["charm"])
+            continue
+
+        # Follow default image-gather process for dependency apps
+        if is_dependency_app(app):
+            logging.info("Dependency app '%s' with charm '%s'", app_name,
+                         app["charm"])
+            images.extend(get_dependency_app_images(app))
+            continue
+
+        # Ensure analytics app has necessary github repo/branch metadata
+        if not bundle_app_contains_gh_metadata(app):
+            raise KeyError(
+                "Application '%s' doesn't include expected gh metadata keys" %
+                app_name
+            )
+
+        images.extend(get_analytics_app_images(app))
+
+    return cleanup_images(images)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Gather all images from a bundle"
+    )
+    parser.add_argument("bundle")
+    parser.add_argument("--airgap-testing", action="store_true",
+                        help="Include images needed for airgap testing.")
+
+    args = parser.parse_args()
+
+    images = get_bundle_images(args.bundle)
+
+    # append the airgap images
+    if args.airgap_testing:
+        images.extend(AIRGAP_TESTING_IMAGES)
+
+    logging.info(f"Found {len(images)} different images")
+
+    for img in images:
+        print(img)
```

Comment:
I still think it is a better system design to make this script save the file in a user specified `--output`. Since this script is potentially going to be used by humans, it'd make a better UX.

Reply:
Why can't humans do `> images.txt`? :)

Reply:
I propose we have `--output` argument but make it optional, and the default being `images.txt`. When I use this, I'm rarely changing the file. Maybe how use it for image scanning is different? Are you frequently generating different files?

Reply:
Folks, I disagree here.

> I still think it is a better system design to make this script save the file in a user specified --output

CLI tools allow you to configure the output format (usually with an `--output` parameter) that controls the structure of the output text. I.e. `kubectl` and `aws` clis both have an `--output` arg for controlling if the output should be yaml, json, text etc. `gcloud` cli has usually `--filter` or `--sort-by`. But none to put into a file.

What the user does then with this output is up to them, whether they pipe it `|` to another program or put it into a file `>`.

Reply:
this conversion is getting too long. I don't fully agree with @kimwnasptd that we should design this the same way `kubectl` is designed, but rather to our own use case. However, I'm ok with stdout since the [CI](https://github.com/canonical/bundle-kubeflow/actions/runs/10708875310) is already implemented to work with this output, and it is well documented in the readme.
@DnPlas I will leave this unresolved for your last review, for me we can resolve it.

Reply:
I've created a follow up issue to agree on the output mechanisms https://github.com/canonical/bundle-kubeflow/issues/1060

For now, since the current functionality is good enough and documented I'll move on and resolve this conversation